### DISCUSSION TITLE
Late Initialization Keyword - For members initialized outside of a constructor

### DISCUSSION THREAD

**Discussion by thomhurst** [dotnet/csharplang#7081] [Language Ideas]
How many times you have seen the message 
`Non-nullable field '_myfield' is uninitialized. Consider declaring the field as nullable.`
when you're confident it's not going to be null?

Take this simple code:

```csharp
public class MyClass
{
    private readonly string _myfield;

    public required string MyProperty
    {
        get => _myfield;
        init => _myfield = value;
    }
}
```

Since `MyProperty` is required, and will set `_myfield` - Then we can deduce that `_myfield` will not be null. But we'll still get the compiler warning.

The current workaround is this:

`private readonly string _myfield = null!;`
or
`private readonly string _myfield = default!;`

Now, I find this a weird syntax. It's pretty ugly and it doesn't really make sense to read.

> "Here's my nullable field, I know it's not null, so I'll set it a value of null but tell the compiler that it's not null really"

It also has the fallback of IF it actually ever is null, it's going to be hard to track errors, and you'll just get a NullReferenceException as normal.

Kotlin has a nice approach to this, a `lateinit` keyword.
C# could of course choose another keyword, but this tells the compiler a similar thing, that the value won't be null because we'll be setting it later on.
But another benefit is that if that value is ever tried to get accessed BEFORE it actually is initialized, then you'll get a specific exception about it not being initialized. This would make debugging code a lot easier, as you'll not be tracking down a NRE, but you know you'll have missed initializing something.

That could potentially look like:

```csharp
public class MyClass
{
    private late readonly string _myfield;

    public required string MyProperty
    {
        get => _myfield;
        init => _myfield = value;
    }
}
```

Thoughts?

Sorry if there's already a discussion on this, I tried searching but couldn't find anything.

**Comment by 333fred** [comment]
This is the supported way of conveying the information:

```cs
using System.Diagnostics.CodeAnalysis;

public class MyClass
{
    private readonly string _myfield; // No warning

    public required string MyProperty
    {
        get => _myfield;
        [MemberNotNull(nameof(_myfield))]
        init => _myfield = value;
    }
}
```

https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQAwAJMCMAdACJQQDmMA9iglAMYpEDC1AJgKYCCMEANgE8UUFAG4AsAChpmAMz4ATLgCyglv1QppAb2m4DuAA4AnKADcICTrhOcI7ajCH4CeAPphBAMyid+7GK4APTBuABy1LgA7hAmMLAU0vqG8racAI4ArlB27K54agAKJtRGnCYIgikGelKGDbgUnAi4ALwAfLiePn4BkvWNBgDaKpxgwBWRCOFZ/PwAFHxgnNTeCz2+/uwAlDsAujWNsFCtnd1eWwHtuJb8WZwDDQC+0s9AA=

**Reply by thomhurst** [reply]
Still doesn't offer you any insight into the fact it hasn't been initialized when you've used it though.

Also it's pretty verbose !

**Reply by AartBluestoke** [reply]
Failure to initialise an init field is an error:

```cs
var mc = new MyClass(); // <-- error CS9035: Required member 'MyClass.MyProperty' must be set in the object initializer or attribute constructor.

public class MyClass
{
    private readonly string _myfield; // No warning

    public required string MyProperty
    {
        get => _myfield;
        [MemberNotNull(nameof(_myfield))]
        init => _myfield = value;
    }
    
}
```
https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQAwAJMCMAdACJQQDmMA9iglAMYpEDC1AJgKYCCMEANgE8UUFAG4AsAChpANwgAnXGAa4AvLhicA7rgCyglv1QoAFAEpJMqZgDM+AEz7DxlCmkBvabh+4ADgpQ8gicuAqcEOzUMEL4BHgA+mCCAGZQnPzsYrgA9Dm4AHLUuNqKMLAU0t6+dmGcAI4ArlDh7HF4BgAKCtR+nAoIgtU+XlK+47gUnAjqAHy4SanpmVYTvgDaepxgwP1FCAWN/PymfGCc1Cmmi2kZ7ObmALrDE7BQM2rzN8ttGvL8jU4q18AF8XtIwdYgA==
